cachier/pickle_core.py

```diff
@@ -131,52 +131,113 @@ def _get_cache(self):
+                try:
+                    res = pickle.load(cache_file)
+                except EOFError:
+                    res = None
+        except FileNotFoundError:
+            res = None
+        return res
+
+    def _clear_all_cache_files(self):
+        fpath = self._cache_fpath()
+        path, name = os.path.split(fpath)
+        for subpath in os.listdir(path):
+            if subpath.startswith(name):
+                os.remove(os.path.join(path, subpath))
+
+    def _clear_being_calculated_all_cache_files(self):
+        fpath = self._cache_fpath()
+        path, name = os.path.split(fpath)
+        for subpath in os.listdir(path):
+            if subpath.startswith(name):
+                entry = self._get_cache_by_key(hash=subpath.split('_')[-1])
+                entry['being_calculated'] = False
+                self._save_cache(entry, hash=subpath.split('_')[-1])
+
+    def _save_cache(self, cache, key=None, hash=None):
         with self.lock:
             self.cache = cache
             fpath = self._cache_fpath()
+            if key is not None:
+                fpath += f'_{hashlib.sha256(pickle.dumps(key)).hexdigest()}'
+            elif hash is not None:
+                fpath += f'_{hash}'
             with portalocker.Lock(fpath, mode='wb') as cache_file:
                 pickle.dump(cache, cache_file, protocol=4)
-            self._reload_cache()
+            if key is None:
+                self._reload_cache()
 
     def get_entry_by_key(self, key, reload=False):  # pylint: disable=W0221
         with self.lock:
-            # print('{}, {}'.format(self.reload, reload))
-            if self.reload or reload:
-                self._reload_cache()
-            return key, self._get_cache().get(key, None)
+            if self.separate_files:
+                return key, self._get_cache_by_key(key)
+            else:
+                if self.reload or reload:
+                    self._reload_cache()
+                return key, self._get_cache().get(key, None)
 
     def get_entry(self, args, kwds, hash_params):
         key = args + tuple(sorted(kwds.items())) if hash_params is None else hash_params(args, kwds)
         # print('key type={}, key={}'.format(type(key), key))
         return self.get_entry_by_key(key)
 
     def set_entry(self, key, func_res):
-        with self.lock:
-            cache = self._get_cache()
-            cache[key] = {
+        key_data = {
                 'value': func_res,
                 'time': datetime.now(),
                 'stale': False,
                 'being_calculated': False,
             }
-            self._save_cache(cache)
+        if self.separate_files:
+            self._save_cache(key_data, key)
+        else:
+            with self.lock:
+                cache = self._get_cache()
+                cache[key] = key_data
+                self._save_cache(cache)
+
+    def mark_entry_being_calculated_separate_files(self, key):
+        self._save_cache({
```

Comment:
Why does the `separate_files` version of `mark_entry_being_calculated` always override any possible existing entry, setting its `value` to `None`, its `time` to to `datetime.now()`  and its `stale` to `False`?

What if a calculation was triggered with `overwrite_cache=True`? There is some existing, possibly non-stale value with a specific time already in the cache. "This" call might wish to overwrite the cache, but what if a call from another thread is triggered while this is running? I'm thinking out loud here. I think this is OK, since one something is marked as `being_calculated`, no other call gets to access the existing value, if even if it's not stale.

Give this a thought, though. Double check me, since you've just delved into the code. :)